### PR TITLE
refactor: refactor apply_cse by using cached expr

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

This PR introduces a new method `run_exprs` and a new field `cached_values` to `Evaluator` to avoid eval the same expression.

After this, we can delete the `apply_cse` to simple codes.

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13104)

### CODE REVIEW

File: src/query/sql/src/evaluator/block_operator.rs [hunk 1]
```diff
@@ -91,15 +91,14 @@ impl BlockOperator {
...
-                        let evaluator = Evaluator::new(&input, func_ctx, &BUILTIN_FUNCTIONS);
-                        let result = evaluator.run(expr)?;
-                        let col = BlockEntry::new(expr.data_type().clone(), result);
-                        input.add_column(col);
-                    }
+                    // Local variable run in single thread, so `Rc` and `RefCell` are enough.
+                    let evaluator =
+                        Evaluator::new(&input, func_ctx, &BUILTIN_FUNCTIONS).with_cache();
+                    let result = evaluator.run_exprs(exprs)?;
+                    let block = DataBlock::new(result, input.num_rows());
```
Comment (sundy-li) [inline src/query/sql/src/evaluator/block_operator.rs:98]: input should add results to build the final block.

**Comment by ariesdevil** [comment]
This PR needs to run benchmark to see if there are any performance regressions

File: src/query/expression/src/evaluator.rs [hunk 2]
```diff
@@ -89,6 +104,25 @@ impl<'a> Evaluator<'a> {
         }
     }
 
+    pub fn run_exprs(&self, exprs: &[Expr]) -> Result<Vec<BlockEntry>> {
+        let mut columns: Vec<BlockEntry> = Vec::with_capacity(exprs.len());
+        for expr in exprs {
+            // try get result from cache
+            if let Some(cached_values) = &self.cached_values {
```
Comment (sundy-li) [inline src/query/expression/src/evaluator.rs:111]: seems this is already covered in L141

File: src/query/expression/src/evaluator.rs [hunk 3]
```diff
@@ -103,6 +137,13 @@ impl<'a> Evaluator<'a> {
         #[cfg(debug_assertions)]
         self.check_expr(expr);
 
+        // try get result from cache
+        if let Some(cached_values) = &self.cached_values {
```
Comment (sundy-li) [inline src/query/expression/src/evaluator.rs:141]: if !matches!(expr, Expr::ColumnRef {..} | Expr::Constant{..}) && let Some(cached_values) = &self.cached_values {

File: src/query/expression/src/evaluator.rs [hunk 4]
```diff
@@ -89,6 +104,16 @@ impl<'a> Evaluator<'a> {
         }
     }
 
+    pub fn run_exprs(&self, exprs: &[Expr]) -> Result<Vec<BlockEntry>> {
```
Comment (sundy-li) [inline src/query/expression/src/evaluator.rs:107]: should return `Result<Vec<Value<AnyType>>>`, keep consistent with `run` function.

Comment (sundy-li) [reply src/query/expression/src/evaluator.rs:107]: seems we don't need this function, better to call `run` function outside.

Comment (ariesdevil) [reply src/query/expression/src/evaluator.rs]: thx, fixed

**Comment by ariesdevil** [comment]
The `Expr::Cast` run in different evaluator lead some issue, reproduce sql:
```sql
create table t(a int);
insert into t values(1);
select sum(case when a < 1 then cast(a as decimal(2, 0)) else 0 end) as a_old, sum(case when a >= 1 then cast(a as decimal(2,0)) else 0 end) as a_new from t;
```
output should be `0, 1`, but got `0, null`

**Review by sundy-li — APPROVED** [review]

**Pull request merged**